.github/workflows/check-build-publish.yaml

```diff
@@ -0,0 +1,106 @@
+name: Check, build and publish
+
+on:
+  push:
+    branches:
+      - main
+
+concurrency: 
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  checks:
+    name: Test and lint
+    runs-on: ubuntu-22.04
+    permissions:
+      checks: write
+    steps:
+      - name: Install build dependencies
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y libssl-dev pkg-config protobuf-compiler libprotobuf-dev
+
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Setup rust toolchain
+        uses: actions-rs/toolchain@v1
+        with: # seems to ignore rust-toolchain but it should be supported
+          toolchain: "1.76.0"
+          components: rustfmt, clippy
+
+      - name: Check code formatting
+        uses: actions-rs/cargo@v1
+        with:
+          command: fmt
+          args: --check
+      
+      - name: Clippy linting
+        uses: giraffate/clippy-action@v1
+        with:
+          reporter: "github-pr-review"
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Dependencies security audit 
+        uses: actions-rust-lang/audit@v1
+        with:
+          ignore: RUSTSEC-2023-0071 # patch not available att
+
+      - name: Run tests
+        run: |
+          cp example.env .env
+          cargo install cargo-make
+          cargo make db-up
+          cargo make full-test
+
+  publish:
+    name: Publish to ghcr.io
+    runs-on: ubuntu-22.04
+    permissions:
+      contents: read
+      packages: write
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v4
+
+    - name: Install build dependencies
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y libssl-dev pkg-config protobuf-compiler libprotobuf-dev
+
+    - name: Setup rust toolchain
+      uses: actions-rs/toolchain@v1
+      with: # seems to ignore rust-toolchain but it should be supported
+        toolchain: "1.76.0"
```

Comment:
To be honest I think they're totally different cases. Ubuntu releases can shake a lot of things up, Rust has a compatibility guarantee that every version is backwards compatible, with very, very, very, very, very few exceptions. Said exceptions address soundness bugs that only affect maybe 0.0001% of code that's doing extremely fancy `unsafe` or lifetime shenanigans, and even then they typically warn for at least an entire release cycle that tells you to fix it before the next one. The only other changes that exist we don't have to worry about like ABI changes (we aren't, and shouldn't, dynamically link Rust code without a `#[repr(c)]` intermediate interface which obviates the issue).

Fixing `Cargo.lock` is reasonable, but I don't think pinning Rust does anything meaningful.

Sure, we could end up auto-updating into a CVE, but just as easily could a current version have a CVE or other problem and we don't notice and update as quickly as we would have as a new point release fixing it. The potential impact of a CVE on the ratings center is pretty minimal on its own anyway, and IMO we're at *far* more risk of an actually scary one being in something like `tokio`, `tonic`, or `hyper` than Rust anyway.

Reply:
To be clear, this does depend on language for me. For Java, PHP, or Python I'd definitely pin. Rust, C, Go, and C++ I would not pin.

Reply:
If we do add `-Dwarnings`, that's the only thing that might conflict with auto-updates (since a new warning being introduced could introduce a compilation failure), I'd be in favor of not doing `-Dwarnings` if that's a concern, but any new warnings that pop up are probably fairly severe.

Reply:
Okay lets give a whirl on stable!